### DISCUSSION TITLE
How to request an object to be executed only between two power iterations?

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#26491] [Q&A General]
How can I set the `execute_on` parameter to ensure that the target object is only executed between two power iterations of InversePowerMethod? It may not be `execute_on=NONLINEAR` because InversePowerMethod might not necessarily involve the concept of `NONLINEAR`. Otherwise, why would the operation of computing the source term |Bx| require the `execute_on` to include `LINEAR`, when it actually only needs to be executed between separate power iterations?

Any assistance will be appreciated.

**Comment by GiudGiud** [comment]
Hello

Seems like we would need a new execute_on flag to do that. 
Can you print a verbose log with the power iterations, the nonlienar iterations, and the execution of MOOSE objects in between?

You can use this to detect execution btw
https://mooseframework.inl.gov/source/userobjects/MessageFromInput.html
or the Debug flags.

Guillaume

**Reply by hityyds** [reply]
Thank you very much for your suggestion. Following your advice, I used the `show_execution_order = 'LINEAR NONLINEAR' `command to output some information about the execution order. I still observe NONLINEAR operations which are next to LINEAR operations in the InversePowerMethod. As I am using an SN program with numerous variables, the output is quite lengthy, making it inconvenient to share here. 

Although I still have one question at the moment, I can explore and test on my own. Great appreciation is owed to you. If you're interested, feel free to continue reading the following content, and I can also inform you about the results of subsequent tests if you wish. 

In each power iteration, Petsc is essentially solving an Ax=b problem. Despite InversePowerMethod being an eigenvalue solver, the eigenvalue solution is independent of Petsc. The vector b contains information from the previous power iteration's eigenvalue and variable solutions. The object I intend to create is an acceleration algorithm, and I want this object to be executed before b is passed into Petsc. According to the documentation, when execute_on=NONLINEAR, the object is executed before Jacobian evaluation. However, I still cannot determine whether it happens before or after b is imported into Petsc.

**Reply by GiudGiud** [reply]
> the object is executed before Jacobian evaluation

both on nonlinear iterations and on power iterations right? So it does not work for you.
I think the way forward for now is to hardcode the execution of the objects in the Eigenvalue executioner.

Maybe @loganharbour or @YaqiWang will see an easier solution to execute on power iterations

**Reply by lindsayad** [reply]
Note that we prefer the `Eigenvalue` executioner over `InversePowerMethod`. The latter will be removed sometime in the future
